rewrite/rewrite/python/format/spaces_visitor.py

```diff
@@ -408,6 +408,41 @@ def visit_expression_type_tree(self, expression_type_tree: ExpressionTypeTree, p
         ett = space_before(ett, False)
         return ett
 
+    def visit_comprehension_expression(self, comprehension_expression: ComprehensionExpression, p: P) -> J:
+        ce = cast(ComprehensionExpression, super().visit_comprehension_expression(comprehension_expression, p))
+
+        # Handle space before result this will depend on the style setting for the comprehension type.
+        if ce.kind == ComprehensionExpression.Kind.LIST:
+            ce = ce.with_result(space_before(ce.result, self._style.within.brackets))
+            ce = ce.with_suffix(update_space(ce.suffix, self._style.within.brackets))
+        elif ce.kind == ComprehensionExpression.Kind.GENERATOR:
+            ce = ce.with_result(space_before(ce.result, False))
+            ce = ce.with_suffix(update_space(ce.suffix, False))
+        elif ce.kind in (ComprehensionExpression.Kind.SET, ComprehensionExpression.Kind.DICT):
+            ce = ce.with_result(space_before(ce.result, self._style.within.braces))
+            ce = ce.with_suffix(update_space(ce.suffix, self._style.within.braces))
+
+        return ce
+
+    def visit_comprehension_condition(self, condition: ComprehensionExpression.Condition, p: P) -> J:
+        cond = cast(ComprehensionExpression.Condition, super().visit_comprehension_condition(condition, p))
+        # Set single space before and after comprehension 'if' keyword.
+        cond = space_before(cond, True)
+        cond = cond.with_expression(space_before(cond.expression, True))
+        return cond
+
+    def visit_comprehension_clause(self, clause: ComprehensionExpression.Clause, p: P) -> J:
+        cc = cast(ComprehensionExpression.Clause, super().visit_comprehension_clause(clause, p))
+
+        # Ensure single space before 'for' keyword
+        cc = space_before(cc, True)
+
+        # Space around iterator variable and after 'in' keyword e.g. for  i  in  range(10) <-> ...for i in range(10)
+        cc = cc.padding.with_iterated_list(space_before_left_padded(cc.padding.iterated_list, True))
+        cc = cc.with_iterator_variable(space_before(cc.iterator_variable, True))
+        cc = cc.padding.with_iterated_list(space_before_left_padded_element(cc.padding.iterated_list, True))
```

Comment:
Can you verify that the first and last of these statements are both required? They look very similar and I don't know what the difference is, really.

Reply:
Yes, I have added some extra comments to their exact purpose. For extra validation, you can also comment on one, which will result in the tests failing.